### PR TITLE
Handling docker resource from CLI.

### PR DESCRIPTION
**Pull request by Veebers**
## Description of change

This adds support in Juju for the docker resource type (in charms).

Currently it is a WIP and needs further work.

## QA steps

Using custom charms, ensure that they deploy correctly (i.e. use the resources as they should be).
Further solid QA steps to come in the lifetime of this WIP.

## Documentation changes

There will need to be some help documentation changes to the deploy command.
There will need to be some help documentation changes to the help regarding charms and resource types.

### CODE REVIEW

**Comment by Veebers** [comment]
This is an initial WIP that will support charms using the docker resource type.

Still need to wire up:
  - Storing the docker details in a collection
  - Once that's done I can wire up the retrieval of those details
  - And once that's done actually have the resource-get command output the docker image details.

File: apiserver/resources.go [hunk 1]
```diff
@@ -144,19 +150,18 @@ func (h *ResourcesHandler) readResource(backend ResourcesBackend, req *http.Requ
...
-		}
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
 
-	ext := path.Ext(res.Path)
-	if path.Ext(uReq.Filename) != ext {
-		return nil, errors.Errorf("incorrect extension on resource upload %q, expected %q", uReq.Filename, ext)
+	if res.Type == charmresource.TypeFile {
```
Comment (wallyworld) [inline apiserver/resources.go]: For docker resources we should validate that the content parses as a valid docker image url

File: apiserver/resources.go [hunk 2]
```diff
@@ -213,7 +218,8 @@ func extractUploadRequest(req *http.Request) (api.UploadRequest, error) {
 		return ur, errors.Annotate(err, "invalid fingerprint")
 	}
 
-	filename, err := extractFilename(req)
+	// Extract out to extract value or filename etc.
+	resourceValue, err := extractFilename(req)
```
Comment (wallyworld) [inline apiserver/resources.go]: This is the filename isn't it? I'm not sure about the rename to resourceValue.
Only file type resources will have the filename attribute in the req header set.
Or is the intent to use resourceValue to hold the docker path?

File: cmd/juju/resource/deploy.go [hunk 3]
```diff
@@ -5,17 +5,18 @@ package resource
...
 	"strings"
 
 	"github.com/juju/errors"
 	charmresource "gopkg.in/juju/charm.v6/resource"
-	"gopkg.in/juju/charmrepo.v3/csclient/params"
 	"gopkg.in/macaroon.v2-unstable"
 	"gopkg.in/yaml.v2"
 
 	"github.com/juju/juju/charmstore"
+	coreresources "github.com/juju/juju/core/resources"
```
Comment (wallyworld) [inline cmd/juju/resource/deploy.go]: no need for an alias unless there's a name clash

File: cmd/juju/resource/deploy.go [hunk 4]
```diff
@@ -141,23 +136,36 @@ func (d deployUploader) upload(resourceValues map[string]string, revisions map[s
...
+		return errors.Annotatef(err, "file for resource %q", name)
+	}
+	if err != nil {
+		return errors.Annotatef(err, "can't read file for resource %q", name)
+	}
+	return nil
+}
+
+func checkDockerDetails(name, value string) error {
 	// for all resources that are docker images, make sure the image string is legit i.e. uri@sha256:sha.
```
Comment (wallyworld) [inline cmd/juju/resource/deploy.go]: In controller config, we have already a func validateCAASOperatorImagePath()
We can export it and move it to core/resources and use it here as well

File: cmd/juju/resource/deploy.go [hunk 5]
```diff
@@ -182,10 +190,10 @@ func (d deployUploader) validateResources() error {
 	return nil
 }
 
-// storeResources returns which resources revisions will need to be retrieved
+// csStoreResources returns which resources revisions will need to be retrieved
 // either as they where explicitly requested by the user for that rev or they
 // weren't provided by the user.
-func (d deployUploader) storeResources(uploads map[string]string, revisions map[string]int) []charmresource.Resource {
+func (d deployUploader) csStoreResources(uploads map[string]string, revisions map[string]int) []charmresource.Resource {
```
Comment (wallyworld) [inline cmd/juju/resource/deploy.go]: why the rename?

File: core/resources/resources.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resources
+
+// DockerDetails holds the details for a Docker resource type.
+type DockerDetails struct {
+	// RegistryName holds the image name (including host and sha256) of the image in the docker registry.
```
Comment (wallyworld) [inline core/resources/resources.go]: RegistryPath. ..... image path

File: core/resources/resources.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resources
+
+// DockerDetails holds the details for a Docker resource type.
+type DockerDetails struct {
```
Comment (wallyworld) [inline core/resources/resources.go]: Maybe call this DockerImageDetails

File: dependencies.tsv [hunk 8]
```diff
@@ -111,7 +111,7 @@ gopkg.in/goose.v2	git	36df5d12fc6d0ef1c102e1c18a22ddf345b18821	2018-03-22T12:54:
 gopkg.in/httprequest.v1	git	1a21782420ea13c3c6fb1d03578f446b3248edb1	2018-03-08T16:26:44Z
 gopkg.in/ini.v1	git	776aa739ce9373377cd16f526cdf06cb4c89b40f	2016-02-22T23:24:41Z
 gopkg.in/juju/blobstore.v2	git	51fa6e26128d74e445c72d3a91af555151cc3654	2016-01-25T02:37:03Z
-gopkg.in/juju/charm.v6	git	12512ca9cd830083aad030b50b1dc2a62725daaa	2018-05-24T01:58:33Z
+gopkg.in/juju/charm.v6	git	0f7685c8057defe2282747a237e119d18fcf8e20	2018-06-27T22:14:11Z
```
Comment (wallyworld) [inline dependencies.tsv]: this isn't the latest commit

File: state/docker_resource.go [hunk 9]
```diff
@@ -8,11 +8,12 @@ import (
 	"io"
 
 	"github.com/juju/errors"
-	"gopkg.in/juju/charmrepo.v3/csclient/params"
 	"gopkg.in/mgo.v2"
 	"gopkg.in/mgo.v2/bson"
 	"gopkg.in/mgo.v2/txn"
 	"gopkg.in/yaml.v2"
+
+	coreresources "github.com/juju/juju/core/resources"
```
Comment (wallyworld) [inline state/docker_resource.go]: ditto about unnecessary alias

File: state/docker_resource.go [hunk 10]
```diff
@@ -24,8 +25,8 @@ type dockerResourceDoc struct {
 	// Id holds the resource ID
 	Id string `bson:"_id"`
 
-	// ImageName holds the image name (including host) of the image in the docker registry.
-	ImageName string `bson:"imagename"`
+	// RegistryPath holds the image name (including host) of the image in the docker registry.
+	RegistryPath string `bson:"registrypath"`
```
Comment (wallyworld) [inline state/docker_resource.go]: we normally hyphenate compound words here

File: state/docker_resource.go [hunk 11]
```diff
@@ -64,13 +65,13 @@ func (dr DockerResource) Save(resourceID string, drInfo params.DockerInfoRespons
...
 			return []txn.Op{{
-				C:      dockerResourceC,
+				C:      dockerResourcesC,
 				Id:     existing.Id,
 				Assert: txn.DocExists,
 				Update: bson.D{
 					{"$set",
 						bson.D{
-							{"imagename", doc.ImageName},
+							{"imagename", doc.RegistryPath},
```
Comment (wallyworld) [inline state/docker_resource.go]: "registry-path"

Hopefully the units tests would have failed here

File: state/docker_resource.go [hunk 12]
```diff
@@ -109,10 +110,10 @@ func (dr DockerResource) Get(resourceID string) (io.ReadCloser, int64, error) {
 		return nil, -1, errors.Trace(err)
 	}
 	data, err := yaml.Marshal(
```
Comment (wallyworld) [inline state/docker_resource.go]: I don't think it's the responsibility of the persistence layer to construct this as a stream here.
Get and Save should be consistent. If a higher layer needs a reader, it should be constructed there.

File: state/resources_state_resource.go [hunk 13]
```diff
@@ -312,7 +331,16 @@ func (st resourceState) OpenResource(applicationID, name string) (resource.Resou
 		return resource.Resource{}, nil, errors.NotFoundf("resource %q", name)
 	}
 
-	resourceReader, resSize, err := st.storage.Get(storagePath)
+	var resourceReader io.ReadCloser
+	var resSize int64
+	switch resourceInfo.Type {
+	case charmresource.TypeDocker:
```
Comment (wallyworld) [inline state/resources_state_resource.go]: Oh, regarding my comment about the reader in state layer - I see that's an existing way we do it, so best to be consistent. So ignore that other comment.

File: state/resources_state_resource.go [hunk 14]
```diff
@@ -276,7 +280,22 @@ func (st resourceState) storeResource(res resource.Resource, r io.Reader) error
...
+		var dockerDetails coreresources.DockerDetails
+		respBuf := new(bytes.Buffer)
+		respBuf.ReadFrom(r)
+		err = yaml.Unmarshal(respBuf.Bytes(), &dockerDetails)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		err = st.dockerMetadataStorage.Save(storagePath, dockerDetails)
```
Comment (wallyworld) [inline state/resources_state_resource.go]: we want to use res.ID here

File: cmd/juju/resource/deploy.go [hunk 15]
```diff
@@ -110,8 +114,19 @@ func (d deployUploader) upload(files map[string]string, revisions map[string]int
...
+		var (
+			id  string
+			err error
+		)
+		switch d.resources[name].Type {
+		case charmresource.TypeFile:
+			id, err = d.uploadFile(name, resValue)
+		case charmresource.TypeDocker:
+			id, err = d.uploadDockerDetails(name, resValue)
+			// default error...
```
Comment (wallyworld) [inline cmd/juju/resource/deploy.go]: There used to be a default error raised here?  Any reason for removing it?

File: apiserver/resources.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/resources.go]: Should call checkDockerDetails() for docker resources here

File: cmd/juju/resource/deploy.go [hunk 16]
```diff
@@ -121,19 +136,39 @@ func (d deployUploader) upload(files map[string]string, revisions map[string]int
...
-		if os.IsNotExist(err) {
-			return errors.Annotatef(err, "file for resource %q", name)
+func (d deployUploader) validateResourceDetails(resources map[string]string) error {
+	for name, value := range resources {
+		var err error
+		switch d.resources[name].Type {
+		case charmresource.TypeFile:
+			err = d.checkFile(name, value)
+		case charmresource.TypeDocker:
+			err = checkDockerDetails(name, value)
```
Comment (wallyworld) [inline cmd/juju/resource/deploy.go]: The file check about actually checks that the file exists, as opposed to a simple syntax check. The docker equivalent should probably be to poke the URL to check for 404 etc. Ok to add a TODO for now but will need to be fixed.

File: state/docker_resource.go [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/mgo.v2/txn"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/core/resources"
+)
+
+// DockerResource implements DockerResourceStorage
+type DockerResource struct {
```
Comment (wallyworld) [inline state/docker_resource.go]: This doesn't need to be exported.
Should probably be called dockerResourceStorage

File: state/docker_resource.go [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// DockerResourceStorage provides the interface for storing Docker resource-type data
+type DockerResourceStorage interface {
+	Save(resourceID string, drInfo resources.DockerImageDetails) error
+	Remove(resourceID string) error
+	Get(resourceID string) (io.ReadCloser, int64, error)
+}
+
+func NewDockerResource(st *State) *DockerResource {
```
Comment (wallyworld) [inline state/docker_resource.go]: Need a comment. Should probably be called NewDockerResourceStorage

File: state/docker_resource.go [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+	return &DockerResource{
+		st: st,
+	}
+}
+
+// Save creates a new record the a Docker resource.
+func (dr DockerResource) Save(resourceID string, drInfo resources.DockerImageDetails) error {
```
Comment (wallyworld) [inline state/docker_resource.go]: We tend to use pointer receivers (more by convention) in state

File: state/docker_resource_test.go [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/core/resources"
+	"github.com/juju/juju/state"
+)
+
+type dockerResourceSuite struct {
+	ConnSuite
+	resource *state.DockerResource
```
Comment (wallyworld) [inline state/docker_resource_test.go]: In line with other comments, call this resourceStorage

File: state/resources_state_resource.go [hunk 21]
```diff
@@ -75,10 +78,11 @@ type resourceStorage interface {
 }
 
 type resourceState struct {
-	persist resourcePersistence
-	raw     rawState
-	storage resourceStorage
-	clock   clock.Clock
+	persist               resourcePersistence
+	raw                   rawState
+	dockerMetadataStorage DockerResourceStorage
```
Comment (wallyworld) [inline state/resources_state_resource.go]: This makes me think DockerMetadataStorage / DockerMetadata etc is a better name for this stuff? Resource is more a Juju specific term but what we are really doing is storing metadata used to access a docker image

**Comment by Veebers** [comment]
$$merge$$

**Comment by Veebers** [comment]
$$merge$$

**Pull request merged**